### PR TITLE
[system-command-line] switched instantiation to separate command

### PR DESCRIPTION
**Pull request by vlada-shubina**
Next step of https://github.com/dotnet/templating/issues/3809

### Problem
when I started to work on tests I noticed that lots of tests are failing and reason for that is legacy options defined on 'NewCommand' level are "swallowing" some of the arguments that should have been passed to `TemplateCommand`.

### Solution
I decided to split the instantiate logic to a separate command with "clean" reparsing beforehand using the command that doesn't have legacy options defined. Same command can be used for `dotnet new create` subcommand if we decide to enable that. Reparsing allows clean set of attributes that can be used to evaluate template to be run without messing with legacy attributes. Also it allows clean separation between non-legacy and legacy logic.

Other changes:
- implemented reparsing for output option when creating `CliTemplateEngineHost` - this was broken. 
- refactoring - split legacy logic for `NewCommand` to separate file.

Help challenges: help is still to be discussed. 
- now we redefine help option to intercept help - not a good approach
- if we keep short name argument and remaining argument on `NewCommand` level they are also shown in subcommand help - not desired behavior
- currently I'm redirecting help to `InstantiateCommand` however had to make a trick of adding subcommands from `NewCommand` to `InstantiateCommand` so that they are shown in help when doing `dotnet new help`. 

Now that this part is solved I will continue migrate instantiate tests (at least those which do not involve invalid parameter handling).

+ added more tests and small fixes

### Checks:
- [x] Added unit tests 
- [x] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

### CODE REVIEW

File: src/Microsoft.TemplateEngine.Cli/Commands/BaseCommand.cs [hunk 1]
```diff
@@ -146,19 +182,8 @@ protected virtual IEnumerable<string> GetSuggestions(TArgs args, IEngineEnvironm
...
-            IEngineEnvironmentSettings environmentSettings = new EngineEnvironmentSettings(
-                new CliTemplateEngineHost(_host, outputPath),
-                settingsLocation: args.DebugCustomSettingsLocation,
-                virtualizeSettings: args.DebugVirtualizeSettings,
-                environment: new CliEnvironment());
-            return environmentSettings;
-        }
-
         protected abstract Task<NewCommandStatus> ExecuteAsync(TArgs args, IEngineEnvironmentSettings environmentSettings, InvocationContext context);
+#pragma warning restore SA1202 // Elements should be ordered by access
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/BaseCommand.cs:186]: fix me

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.cs [hunk 2]
```diff
@@ -5,42 +5,277 @@
...
+                    hostSpecificDataLoader,
+                    TelemetryLogger);
+
+                return await templateListCoordinator.DisplayCommandDescriptionAsync(instantiateArgs, default).ConfigureAwait(false);
+            }
+
+            var templates = await templatePackageManager.GetTemplatesAsync(context.GetCancellationToken()).ConfigureAwait(false);
+            var templateGroups = TemplateGroup.FromTemplateList(CliTemplateInfo.FromTemplateInfo(templates, hostSpecificDataLoader));
+
+            //TODO: decide what to do if there are more than 1 group.
```
Comment (DavidKarlas) [inline src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.cs:136]: I believe this should be solved elsewhere, when TemplateGroups are created... it should be checked for conflict and in that case have `ShortName`=`GroundIndentifier`, ugly but whole code base doesn't need to care about this then(like --list)...

**Review by DavidKarlas — APPROVED** [review]
Maybe also add test for `dotnet new create create` and `dotnet new create list` cases.

File: src/Microsoft.TemplateEngine.Cli/Commands/BaseCommand.cs [hunk 3]
```diff
@@ -50,37 +64,59 @@ protected BaseCommand(string name, string? description = null) : base(name, desc
...
+        internal ITelemetryLogger TelemetryLogger { get; }
+
+        internal NewCommandCallbacks Callbacks { get; }
+
+        protected IEngineEnvironmentSettings CreateEnvironmentSettings(GlobalArgs args, ParseResult parseResult)
+        {
+            //reparse to get output option if present
+            //it's kept private so it is not reused for any other purpose except initializing host
+            //for template instantiaton it has to be reparsed
+            string? outputPath = ParseOutputOption(parseResult);
```
[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/Commands/BaseCommand.cs:77]: I think the only way to do it properly is do a feature in template engine core instead, allowing to send parameters sent to template instantiation to allow host to populate additional parameters before template instantiation is started. At that time output dir is already known.

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.cs (hunk 2 above)
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.cs:136]: `list` is not really impacted by it, it is possible to show template groups with the same name. But, in general I agree.
It also depends on how we would like to solve it:
- if we want to prefix conflicted short name with group identity: `<group identity>:<short-name>`
- or require it as additional parameter: `<short-name> --group <group identity>`

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.cs [hunk 4]
```diff
@@ -5,42 +5,277 @@
...
+using Microsoft.TemplateEngine.Edge.Settings;
 
 namespace Microsoft.TemplateEngine.Cli.Commands
 {
     internal class InstantiateCommand : BaseCommand<InstantiateCommandArgs>
     {
-        internal InstantiateCommand(ITemplateEngineHost host, ITelemetryLogger logger, NewCommandCallbacks callbacks) : base(host, logger, callbacks, "create")
+        private NewCommand? _parentCommand;
+
+        internal InstantiateCommand(ITemplateEngineHost host, ITelemetryLogger logger, NewCommandCallbacks callbacks) : base(host, logger, callbacks, "create", "TODO")
```
[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.cs:20]: I'm listing TODOs in https://github.com/dotnet/templating/issues/3809 or https://github.com/dotnet/templating/issues/2191, however this particular I forgot, thanks.

I will check all localization / "TODO" in the new code anyway as part of final cleanup  https://github.com/dotnet/templating/issues/3809

**Pull request merged**